Test failures with Python 3.15.0a5: test_date_today + DeprecationWarnings

### Python Version

3.15.0a5

### pytest Version

9.0.2

### Package Version

3.2.0 from main

### Description

Hello. Just a heads up that running `tox -e py315` fails with Python 3.15.0a5.

There is 1 actual failure, and 3 other tests fail due to the same DeprecationWarning, possibly leaking from pytest, not sure.

This is the actual failure:

```
_______________________________ test_date_today ________________________________

    def test_date_today():
        with time_machine.travel(EPOCH):
            today = dt.date.today()
>           assert today.year == 1970
E           assert 2026 == 1970
E            +  where 2026 = datetime.date(2026, 2, 9).year

tests/test_time_machine.py:133: AssertionError
```

@befeleme says it looks like this change in Python could be the culprit: https://github.com/python/cpython/commit/fc3e22a06c9b7c93ee379b5554a40a275745e15f

Full log:

<details>

```
py315: uv-sync> uv sync --locked --python-preference system --extra cli --no-default-groups --no-editable --reinstall-package time-machine --group test -p cpython3.15
py315: commands[0]> python -W error -W ignore:datetime.datetime.utcfromtimestamp:DeprecationWarning -W ignore:datetime.datetime.utcnow:DeprecationWarning -m coverage run -m pytest tests
============================= test session starts ==============================
platform linux -- Python 3.15.0a5, pytest-9.0.2, pluggy-1.6.0
cachedir: .tox/py315/.pytest_cache
Using --randomly-seed=4217544912
rootdir: .../time-machine
configfile: pyproject.toml
plugins: randomly-4.0.1, time-machine-3.2.0
collected 207 items

tests/test_time_machine.py .s...............FF.........................F [ 21%]
...F....F............................................................... [ 56%]
..........................................                               [ 76%]
tests/test_cli.py ................................................       [100%]

=================================== FAILURES ===================================
__________________________ test_async_context_manager __________________________

main = <coroutine object test_async_context_manager.<locals>.record_time at 0x7f84c48e45a0>

    def run(main, *, debug=None, loop_factory=None):
        """Execute the coroutine and return the result.
    
        This function runs the passed coroutine, taking care of
        managing the asyncio event loop, finalizing asynchronous
        generators and closing the default executor.
    
        This function cannot be called when another asyncio event loop is
        running in the same thread.
    
        If debug is True, the event loop will be run in debug mode.
        If loop_factory is passed, it is used for new event loop creation.
    
        This function always creates a new event loop and closes it at the end.
        It should be used as a main entry point for asyncio programs, and should
        ideally only be called once.
    
        The executor is given a timeout duration of 5 minutes to shutdown.
        If the executor hasn't finished within that duration, a warning is
        emitted and the executor is closed.
    
        Example:
    
            async def main():
                await asyncio.sleep(1)
                print('hello')
    
            asyncio.run(main())
        """
        if events._get_running_loop() is not None:
            # fail fast with short traceback
            raise RuntimeError(
                "asyncio.run() cannot be called from a running event loop")
    
        with Runner(debug=debug, loop_factory=loop_factory) as runner:
>           return runner.run(main)
                   ^^^^^^^^^^^^^^^^

/usr/lib64/python3.15/asyncio/runners.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:109: in run
    task = self._loop.create_task(coro, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning

During handling of the above exception, another exception occurred:

    def test_async_context_manager():
        recorded_time = None
    
        async def record_time() -> None:
            nonlocal recorded_time
            async with time_machine.travel(EPOCH + 150.0):
                recorded_time = time.time()
    
>       asyncio.run(record_time())

tests/test_time_machine.py:651: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:203: in run
    with Runner(debug=debug, loop_factory=loop_factory) as runner:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/runners.py:63: in __exit__
    self.close()
/usr/lib64/python3.15/asyncio/runners.py:72: in close
    loop.run_until_complete(loop.shutdown_asyncgens())
/usr/lib64/python3.15/asyncio/base_events.py:698: in run_until_complete
    future = tasks.ensure_future(future, loop=self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/tasks.py:732: in ensure_future
    return loop.create_task(coro_or_future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning
__________________ test_async_context_manager_stops_properly ___________________

main = <coroutine object test_async_context_manager_stops_properly.<locals>.record_times at 0x7f84c4ce39b0>

    def run(main, *, debug=None, loop_factory=None):
        """Execute the coroutine and return the result.
    
        This function runs the passed coroutine, taking care of
        managing the asyncio event loop, finalizing asynchronous
        generators and closing the default executor.
    
        This function cannot be called when another asyncio event loop is
        running in the same thread.
    
        If debug is True, the event loop will be run in debug mode.
        If loop_factory is passed, it is used for new event loop creation.
    
        This function always creates a new event loop and closes it at the end.
        It should be used as a main entry point for asyncio programs, and should
        ideally only be called once.
    
        The executor is given a timeout duration of 5 minutes to shutdown.
        If the executor hasn't finished within that duration, a warning is
        emitted and the executor is closed.
    
        Example:
    
            async def main():
                await asyncio.sleep(1)
                print('hello')
    
            asyncio.run(main())
        """
        if events._get_running_loop() is not None:
            # fail fast with short traceback
            raise RuntimeError(
                "asyncio.run() cannot be called from a running event loop")
    
        with Runner(debug=debug, loop_factory=loop_factory) as runner:
>           return runner.run(main)
                   ^^^^^^^^^^^^^^^^

/usr/lib64/python3.15/asyncio/runners.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:109: in run
    task = self._loop.create_task(coro, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning

During handling of the above exception, another exception occurred:

    def test_async_context_manager_stops_properly():
        recorded_times = []
    
        async def record_times() -> None:
            recorded_times.append(time.time())
    
            async with time_machine.travel(EPOCH + 160.0):
                recorded_times.append(time.time())
    
            recorded_times.append(time.time())
    
>       asyncio.run(record_times())

tests/test_time_machine.py:667: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:203: in run
    with Runner(debug=debug, loop_factory=loop_factory) as runner:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/runners.py:63: in __exit__
    self.close()
/usr/lib64/python3.15/asyncio/runners.py:72: in close
    loop.run_until_complete(loop.shutdown_asyncgens())
/usr/lib64/python3.15/asyncio/base_events.py:698: in run_until_complete
    future = tasks.ensure_future(future, loop=self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/tasks.py:732: in ensure_future
    return loop.create_task(coro_or_future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning
_______________________________ test_date_today ________________________________

    def test_date_today():
        with time_machine.travel(EPOCH):
            today = dt.date.today()
>           assert today.year == 1970
E           assert 2026 == 1970
E            +  where 2026 = datetime.date(2026, 2, 9).year

tests/test_time_machine.py:133: AssertionError
_____________________ test_async_context_manager_traveller _____________________

main = <coroutine object test_async_context_manager_traveller.<locals>.test_traveller at 0x7f84c4382f90>

    def run(main, *, debug=None, loop_factory=None):
        """Execute the coroutine and return the result.
    
        This function runs the passed coroutine, taking care of
        managing the asyncio event loop, finalizing asynchronous
        generators and closing the default executor.
    
        This function cannot be called when another asyncio event loop is
        running in the same thread.
    
        If debug is True, the event loop will be run in debug mode.
        If loop_factory is passed, it is used for new event loop creation.
    
        This function always creates a new event loop and closes it at the end.
        It should be used as a main entry point for asyncio programs, and should
        ideally only be called once.
    
        The executor is given a timeout duration of 5 minutes to shutdown.
        If the executor hasn't finished within that duration, a warning is
        emitted and the executor is closed.
    
        Example:
    
            async def main():
                await asyncio.sleep(1)
                print('hello')
    
            asyncio.run(main())
        """
        if events._get_running_loop() is not None:
            # fail fast with short traceback
            raise RuntimeError(
                "asyncio.run() cannot be called from a running event loop")
    
        with Runner(debug=debug, loop_factory=loop_factory) as runner:
>           return runner.run(main)
                   ^^^^^^^^^^^^^^^^

/usr/lib64/python3.15/asyncio/runners.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:109: in run
    task = self._loop.create_task(coro, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning

During handling of the above exception, another exception occurred:

    def test_async_context_manager_traveller():
        recorded_time = None
        shifted_time = None
    
        async def test_traveller() -> None:
            nonlocal recorded_time, shifted_time
            async with time_machine.travel(EPOCH + 170.0, tick=False) as traveller:
                recorded_time = time.time()
                traveller.shift(10.0)
                shifted_time = time.time()
    
>       asyncio.run(test_traveller())

tests/test_time_machine.py:685: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:203: in run
    with Runner(debug=debug, loop_factory=loop_factory) as runner:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/runners.py:63: in __exit__
    self.close()
/usr/lib64/python3.15/asyncio/runners.py:72: in close
    loop.run_until_complete(loop.shutdown_asyncgens())
/usr/lib64/python3.15/asyncio/base_events.py:698: in run_until_complete
    future = tasks.ensure_future(future, loop=self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/tasks.py:732: in ensure_future
    return loop.create_task(coro_or_future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning
___________________________ test_coroutine_decorator ___________________________

main = <coroutine object test_coroutine_decorator.<locals>.record_time at 0x7f84c4383bf0>

    def run(main, *, debug=None, loop_factory=None):
        """Execute the coroutine and return the result.
    
        This function runs the passed coroutine, taking care of
        managing the asyncio event loop, finalizing asynchronous
        generators and closing the default executor.
    
        This function cannot be called when another asyncio event loop is
        running in the same thread.
    
        If debug is True, the event loop will be run in debug mode.
        If loop_factory is passed, it is used for new event loop creation.
    
        This function always creates a new event loop and closes it at the end.
        It should be used as a main entry point for asyncio programs, and should
        ideally only be called once.
    
        The executor is given a timeout duration of 5 minutes to shutdown.
        If the executor hasn't finished within that duration, a warning is
        emitted and the executor is closed.
    
        Example:
    
            async def main():
                await asyncio.sleep(1)
                print('hello')
    
            asyncio.run(main())
        """
        if events._get_running_loop() is not None:
            # fail fast with short traceback
            raise RuntimeError(
                "asyncio.run() cannot be called from a running event loop")
    
        with Runner(debug=debug, loop_factory=loop_factory) as runner:
>           return runner.run(main)
                   ^^^^^^^^^^^^^^^^

/usr/lib64/python3.15/asyncio/runners.py:204: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:109: in run
    task = self._loop.create_task(coro, context=context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning

During handling of the above exception, another exception occurred:

    def test_coroutine_decorator():
        recorded_time = None
    
        @time_machine.travel(EPOCH + 140.0)
        async def record_time() -> None:
            nonlocal recorded_time
            recorded_time = time.time()
    
>       asyncio.run(record_time())

tests/test_time_machine.py:638: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.15/asyncio/runners.py:203: in run
    with Runner(debug=debug, loop_factory=loop_factory) as runner:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/runners.py:63: in __exit__
    self.close()
/usr/lib64/python3.15/asyncio/runners.py:72: in close
    loop.run_until_complete(loop.shutdown_asyncgens())
/usr/lib64/python3.15/asyncio/base_events.py:698: in run_until_complete
    future = tasks.ensure_future(future, loop=self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/tasks.py:732: in ensure_future
    return loop.create_task(coro_or_future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/asyncio/base_events.py:470: in create_task
    task = tasks.Task(coro, loop=self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:268: in extract_stack
    stack = StackSummary.extract(walk_stack(f), limit=limit)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/traceback.py:467: in extract
    return klass._extract_from_extended_frame_gen(
/usr/lib64/python3.15/traceback.py:499: in _extract_from_extended_frame_gen
    linecache.lazycache(filename, f.f_globals)
/usr/lib64/python3.15/linecache.py:217: in lazycache
    lazy_entry = _make_lazycache_entry(filename, module_globals)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.15/linecache.py:238: in _make_lazycache_entry
    loader = _bless_my_loader(module_globals)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

module_globals = {'__builtins__': <module 'builtins' (built-in)>, '__doc__': 'The pytest entry point.', '__file__': '...hon3.15/site-packages/pytest/__main__.py', '__loader__': <coverage.execfile.DummyLoader object at 0x7f84c5721500>, ...}

    def _bless_my_loader(module_globals):
        # Similar to _bless_my_loader() in importlib._bootstrap_external,
        # but always emits warnings instead of errors.
        loader = module_globals.get('__loader__')
        if loader is None and '__spec__' not in module_globals:
            return None
        spec = module_globals.get('__spec__')
    
        # The __main__ module has __spec__ = None.
        if spec is None and module_globals.get('__name__') == '__main__':
            return loader
    
        spec_loader = getattr(spec, 'loader', None)
        if spec_loader is None:
            import warnings
            warnings.warn(
                'Module globals is missing a __spec__.loader',
                DeprecationWarning)
            return loader
    
        assert spec_loader is not None
        if loader is not None and loader != spec_loader:
            import warnings
>           warnings.warn(
                'Module globals; __loader__ != __spec__.loader',
                DeprecationWarning)
E           DeprecationWarning: Module globals; __loader__ != __spec__.loader

/usr/lib64/python3.15/linecache.py:273: DeprecationWarning
=========================== short test summary info ============================
FAILED tests/test_time_machine.py::test_async_context_manager - DeprecationWa...
FAILED tests/test_time_machine.py::test_async_context_manager_stops_properly
FAILED tests/test_time_machine.py::test_date_today - assert 2026 == 1970
FAILED tests/test_time_machine.py::test_async_context_manager_traveller - Dep...
FAILED tests/test_time_machine.py::test_coroutine_decorator - DeprecationWarn...
=================== 5 failed, 201 passed, 1 skipped in 1.49s ===================
py315: exit 1 (2.01 seconds) .../time-machine> python -W error -W ignore:datetime.datetime.utcfromtimestamp:DeprecationWarning -W ignore:datetime.datetime.utcnow:DeprecationWarning -m coverage run -m pytest tests pid=3690691
  py315: FAIL code 1 (2.65=setup[0.64]+cmd[2.01] seconds)
  evaluation failed :( (2.69 seconds)
```

</details>

--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -2,6 +2,14 @@
 Changelog
 =========
 
+* Mock ``datetime.date.today()`` directly, for Python 3.15 support.
+
+  Previously time-machine was mocked only indirectly, since CPython implemented it by calling ``cls.fromtimestamp(time.time())``.
+  Python 3.15 added a fast path that reads the system clock directly (`CPython Issue #130980 <https://github.com/python/cpython/pull/130980>`__), so time travel no longer affected it, which the new mock fixes.
+  Consequently, there are new :ref:`escape hatch <escape-hatch>` functions: ``escape_hatch.datetime.date.today()`` and ``escape_hatch.datetime.datetime.today()``.
+
+  Thanks to Miro Hrončok and Karolina Surma for the report in `Issue #610 <https://github.com/adamchainz/time-machine/issues/610>`__, Lumír 'Frenzy' Balhar for the fix in `PR #618 <https://github.com/adamchainz/time-machine/pull/618>`__, and Maurycy Pawłowski-Wieroński for review
+
 * Build with frame pointers enabled, preparation for `PEP 831 <https://peps.python.org/pep-0831/>`__.
 
   `PR #627 <https://github.com/adamchainz/time-machine/issues/627>`__.

--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -373,9 +373,17 @@ The functions are:
 
   Returns ``True`` if ``time_machine.travel()`` is active, ``False`` otherwise.
 
+* ``escape_hatch.datetime.date.today()``
+
+  Wraps the real ``datetime.date.today()``.
+
 * ``escape_hatch.datetime.datetime.now()``
 
   Wraps the real ``datetime.datetime.now()``.
+
+* ``escape_hatch.datetime.datetime.today()``
+
+  Wraps the real ``datetime.datetime.today()``.
 
 * ``escape_hatch.datetime.datetime.utcnow()``
 

--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -357,6 +357,8 @@ Main API
           with time_machine.travel(...):
               ...
 
+.. _escape-hatch:
+
 Escape hatch API
 ================
 

--- a/docs/usage.rst
+++ b/docs/usage.rst
@@ -89,7 +89,9 @@ Main API
 
   All datetime functions in the standard library are mocked to move to the destination current datetime:
 
+  * ``datetime.date.today()``
   * ``datetime.datetime.now()``
+  * ``datetime.datetime.today()``
   * ``datetime.datetime.utcnow()``
   * ``time.clock_gettime()`` (only for ``CLOCK_REALTIME``)
   * ``time.clock_gettime_ns()`` (only for ``CLOCK_REALTIME``)

--- a/src/_time_machine.c
+++ b/src/_time_machine.c
@@ -8,8 +8,10 @@ typedef struct {
     PyObject *datetime_module;
     PyObject *time_module;
     PyObject *datetime_class;
+    PyObject *date_class;
     PyCFunctionObject *datetime_datetime_now;
     PyCFunctionObject *datetime_datetime_utcnow;
+    PyCFunctionObject *date_today;
     PyCFunctionObject *time_clock_gettime;
     PyCFunctionObject *time_clock_gettime_ns;
     PyCFunctionObject *time_gmtime;
@@ -24,6 +26,8 @@ typedef struct {
     _PyCFunctionFastWithKeywords original_now;
 #endif
     PyCFunction original_utcnow;
+    PyCFunction original_date_today;
+    PyCFunction original_datetime_today;
     PyCFunction original_clock_gettime;
     PyCFunction original_clock_gettime_ns;
     PyCFunction original_gmtime;
@@ -129,6 +133,96 @@ PyDoc_STRVAR(original_utcnow_doc,
     "original_utcnow() -> datetime\n\
 \n\
 Call datetime.datetime.utcnow() after patching.");
+
+/* datetime.date.today() and datetime.datetime.today()
+ * Note: datetime.datetime doesn't define its own today(), it inherits from date.
+ * So we patch date.today() with a wrapper that checks the calling class type.
+ */
+
+static PyObject *
+_time_machine_today(PyObject *cls, PyObject *args)
+{
+    PyObject *time_machine_module = PyImport_ImportModule("time_machine");
+    if (time_machine_module == NULL) {
+        return NULL;  // Propagate ImportError
+    }
+    PyObject *datetime_module = PyImport_ImportModule("datetime");
+    if (datetime_module == NULL) {
+        Py_DECREF(time_machine_module);
+        return NULL;  // Propagate ImportError
+    }
+    PyObject *datetime_class = PyObject_GetAttrString(datetime_module, "datetime");
+    if (datetime_class == NULL) {
+        Py_DECREF(datetime_module);
+        Py_DECREF(time_machine_module);
+        return NULL;  // Propagate AttributeError
+    }
+
+    /* Check if cls is datetime.datetime (or subclass) vs just date */
+    int is_datetime = PyObject_IsSubclass(cls, datetime_class);
+    if (is_datetime == -1) {
+        Py_DECREF(datetime_class);
+        Py_DECREF(datetime_module);
+        Py_DECREF(time_machine_module);
+        return NULL;  // Propagate error
+    }
+
+    const char *func_name = (is_datetime == 1) ? "datetime_today" : "date_today";
+    PyObject *time_machine_func = PyObject_GetAttrString(time_machine_module, func_name);
+    if (time_machine_func == NULL) {
+        Py_DECREF(datetime_class);
+        Py_DECREF(datetime_module);
+        Py_DECREF(time_machine_module);
+        return NULL;  // Propagate AttributeError
+    }
+
+    PyObject *result = PyObject_CallObject(time_machine_func, args);
+
+    Py_DECREF(time_machine_func);
+    Py_DECREF(datetime_class);
+    Py_DECREF(datetime_module);
+    Py_DECREF(time_machine_module);
+
+    return result;
+}
+
+static PyObject *
+_time_machine_original_date_today(PyObject *module, PyObject *args)
+{
+    _time_machine_state *state = get_time_machine_state(module);
+
+    if (state->original_date_today == NULL) {
+        PyErr_SetString(PyExc_ValueError, "Not currently time-travelling.");
+        return NULL;
+    }
+
+    PyObject *result = state->original_date_today(state->date_class, args);
+
+    return result;
+}
+PyDoc_STRVAR(original_date_today_doc,
+    "original_date_today() -> date\n\
+\n\
+Call datetime.date.today() after patching.");
+
+static PyObject *
+_time_machine_original_datetime_today(PyObject *module, PyObject *args)
+{
+    _time_machine_state *state = get_time_machine_state(module);
+
+    if (state->original_datetime_today == NULL) {
+        PyErr_SetString(PyExc_ValueError, "Not currently time-travelling.");
+        return NULL;
+    }
+
+    PyObject *result = state->original_datetime_today(state->datetime_class, args);
+
+    return result;
+}
+PyDoc_STRVAR(original_datetime_today_doc,
+    "original_datetime_today() -> datetime\n\
+\n\
+Call datetime.datetime.today() after patching.");
 
 /* time.clock_gettime() */
 
@@ -458,6 +552,15 @@ _time_machine_patch(PyObject *module, PyObject *unused)
     state->original_utcnow = state->datetime_datetime_utcnow->m_ml->ml_meth;
     state->datetime_datetime_utcnow->m_ml->ml_meth = _time_machine_utcnow;
 
+    /* Patch today() - since datetime.datetime.today() inherits from date.today(),
+     * we only need to patch date.today() once. Save both originals first for
+     * the escape hatch, then patch with a unified wrapper that dispatches based
+     * on the calling class type.
+     */
+    state->original_date_today = state->date_today->m_ml->ml_meth;
+    state->original_datetime_today = state->date_today->m_ml->ml_meth;
+    state->date_today->m_ml->ml_meth = _time_machine_today;
+
     /*
         time.clock_gettime(), only available on Unix platforms.
     */
@@ -517,6 +620,10 @@ _time_machine_unpatch(PyObject *module, PyObject *unused)
     state->datetime_datetime_utcnow->m_ml->ml_meth = state->original_utcnow;
     state->original_utcnow = NULL;
 
+    state->date_today->m_ml->ml_meth = state->original_date_today;
+    state->original_date_today = NULL;
+    state->original_datetime_today = NULL;
+
     /*
         time.clock_gettime(), only available on Unix platforms.
     */
@@ -566,6 +673,14 @@ static PyMethodDef module_functions[] = {
         (PyCFunction)_time_machine_original_utcnow,
         METH_NOARGS,
         original_utcnow_doc},
+    {"original_date_today",
+        (PyCFunction)_time_machine_original_date_today,
+        METH_NOARGS,
+        original_date_today_doc},
+    {"original_datetime_today",
+        (PyCFunction)_time_machine_original_datetime_today,
+        METH_NOARGS,
+        original_datetime_today_doc},
 #if PY_VERSION_HEX >= 0x030d00a2
     {"original_clock_gettime",
         (PyCFunction)_time_machine_original_clock_gettime,
@@ -637,6 +752,17 @@ _time_machine_exec(PyObject *module)
         goto error;
     }
 
+    state->date_class = PyObject_GetAttrString(state->datetime_module, "date");
+    if (state->date_class == NULL) {
+        goto error;
+    }
+
+    state->date_today =
+        (PyCFunctionObject *)PyObject_GetAttrString(state->date_class, "today");
+    if (state->date_today == NULL) {
+        goto error;
+    }
+
     state->time_module = PyImport_ImportModule("time");
     if (state->time_module == NULL) {
         goto error;
@@ -700,8 +826,10 @@ _time_machine_exec(PyObject *module)
 error:
     Py_CLEAR(state->datetime_module);
     Py_CLEAR(state->datetime_class);
+    Py_CLEAR(state->date_class);
     Py_CLEAR(state->datetime_datetime_now);
     Py_CLEAR(state->datetime_datetime_utcnow);
+    Py_CLEAR(state->date_today);
     Py_CLEAR(state->time_module);
     Py_CLEAR(state->time_clock_gettime);
     Py_CLEAR(state->time_clock_gettime_ns);
@@ -719,8 +847,10 @@ _time_machine_traverse(PyObject *module, visitproc visit, void *arg)
     _time_machine_state *state = get_time_machine_state(module);
     Py_VISIT(state->datetime_module);
     Py_VISIT(state->datetime_class);
+    Py_VISIT(state->date_class);
     Py_VISIT(state->datetime_datetime_now);
     Py_VISIT(state->datetime_datetime_utcnow);
+    Py_VISIT(state->date_today);
     Py_VISIT(state->time_module);
     Py_VISIT(state->time_clock_gettime);
     Py_VISIT(state->time_clock_gettime_ns);
@@ -738,8 +868,10 @@ _time_machine_clear(PyObject *module)
     _time_machine_state *state = get_time_machine_state(module);
     Py_CLEAR(state->datetime_module);
     Py_CLEAR(state->datetime_class);
+    Py_CLEAR(state->date_class);
     Py_CLEAR(state->datetime_datetime_now);
     Py_CLEAR(state->datetime_datetime_utcnow);
+    Py_CLEAR(state->date_today);
     Py_CLEAR(state->time_module);
     Py_CLEAR(state->time_clock_gettime);
     Py_CLEAR(state->time_clock_gettime_ns);

--- a/src/_time_machine.c
+++ b/src/_time_machine.c
@@ -8,7 +8,6 @@ typedef struct {
     PyObject *datetime_module;
     PyObject *time_module;
     PyObject *datetime_class;
-    PyObject *date_class;
     PyCFunctionObject *datetime_datetime_now;
     PyCFunctionObject *datetime_datetime_utcnow;
     PyCFunctionObject *date_today;
@@ -27,7 +26,6 @@ typedef struct {
 #endif
     PyCFunction original_utcnow;
     PyCFunction original_date_today;
-    PyCFunction original_datetime_today;
     PyCFunction original_clock_gettime;
     PyCFunction original_clock_gettime_ns;
     PyCFunction original_gmtime;
@@ -136,7 +134,8 @@ Call datetime.datetime.utcnow() after patching.");
 
 /* datetime.date.today() and datetime.datetime.today()
  * Note: datetime.datetime doesn't define its own today(), it inherits from date.
- * So we patch date.today() with a wrapper that checks the calling class type.
+ * So we patch date.today() with a wrapper that calls cls.fromtimestamp(), which
+ * returns the right type for date, datetime, and subclasses of either.
  */
 
 static PyObject *
@@ -146,83 +145,19 @@ _time_machine_today(PyObject *cls, PyObject *args)
     if (time_machine_module == NULL) {
         return NULL;  // Propagate ImportError
     }
-    PyObject *datetime_module = PyImport_ImportModule("datetime");
-    if (datetime_module == NULL) {
-        Py_DECREF(time_machine_module);
-        return NULL;  // Propagate ImportError
-    }
-    PyObject *datetime_class = PyObject_GetAttrString(datetime_module, "datetime");
-    if (datetime_class == NULL) {
-        Py_DECREF(datetime_module);
-        Py_DECREF(time_machine_module);
-        return NULL;  // Propagate AttributeError
-    }
 
-    /* Check if cls is datetime.datetime (or subclass) vs just date */
-    int is_datetime = PyObject_IsSubclass(cls, datetime_class);
-    if (is_datetime == -1) {
-        Py_DECREF(datetime_class);
-        Py_DECREF(datetime_module);
-        Py_DECREF(time_machine_module);
-        return NULL;  // Propagate error
-    }
-
-    const char *func_name = (is_datetime == 1) ? "datetime_today" : "date_today";
-    PyObject *time_machine_func = PyObject_GetAttrString(time_machine_module, func_name);
-    if (time_machine_func == NULL) {
-        Py_DECREF(datetime_class);
-        Py_DECREF(datetime_module);
-        Py_DECREF(time_machine_module);
-        return NULL;  // Propagate AttributeError
-    }
-
-    PyObject *result = PyObject_CallObject(time_machine_func, args);
-
-    Py_DECREF(time_machine_func);
-    Py_DECREF(datetime_class);
-    Py_DECREF(datetime_module);
+    PyObject *timestamp = PyObject_CallMethod(time_machine_module, "time", NULL);
     Py_DECREF(time_machine_module);
-
-    return result;
-}
-
-static PyObject *
-_time_machine_original_date_today(PyObject *module, PyObject *args)
-{
-    _time_machine_state *state = get_time_machine_state(module);
-
-    if (state->original_date_today == NULL) {
-        PyErr_SetString(PyExc_ValueError, "Not currently time-travelling.");
+    if (timestamp == NULL) {
         return NULL;
     }
 
-    PyObject *result = state->original_date_today(state->date_class, args);
+    PyObject *result = PyObject_CallMethod(cls, "fromtimestamp", "O", timestamp);
+
+    Py_DECREF(timestamp);
 
     return result;
 }
-PyDoc_STRVAR(original_date_today_doc,
-    "original_date_today() -> date\n\
-\n\
-Call datetime.date.today() after patching.");
-
-static PyObject *
-_time_machine_original_datetime_today(PyObject *module, PyObject *args)
-{
-    _time_machine_state *state = get_time_machine_state(module);
-
-    if (state->original_datetime_today == NULL) {
-        PyErr_SetString(PyExc_ValueError, "Not currently time-travelling.");
-        return NULL;
-    }
-
-    PyObject *result = state->original_datetime_today(state->datetime_class, args);
-
-    return result;
-}
-PyDoc_STRVAR(original_datetime_today_doc,
-    "original_datetime_today() -> datetime\n\
-\n\
-Call datetime.datetime.today() after patching.");
 
 /* time.clock_gettime() */
 
@@ -540,6 +475,9 @@ _time_machine_patch(PyObject *module, PyObject *unused)
     if (state->original_time)
         Py_RETURN_NONE;
 
+    state->original_date_today = state->date_today->m_ml->ml_meth;
+    state->date_today->m_ml->ml_meth = _time_machine_today;
+
 #if PY_VERSION_HEX >= 0x030d00a4
     state->original_now =
         (PyCFunctionFastWithKeywords)state->datetime_datetime_now->m_ml->ml_meth;
@@ -551,15 +489,6 @@ _time_machine_patch(PyObject *module, PyObject *unused)
 
     state->original_utcnow = state->datetime_datetime_utcnow->m_ml->ml_meth;
     state->datetime_datetime_utcnow->m_ml->ml_meth = _time_machine_utcnow;
-
-    /* Patch today() - since datetime.datetime.today() inherits from date.today(),
-     * we only need to patch date.today() once. Save both originals first for
-     * the escape hatch, then patch with a unified wrapper that dispatches based
-     * on the calling class type.
-     */
-    state->original_date_today = state->date_today->m_ml->ml_meth;
-    state->original_datetime_today = state->date_today->m_ml->ml_meth;
-    state->date_today->m_ml->ml_meth = _time_machine_today;
 
     /*
         time.clock_gettime(), only available on Unix platforms.
@@ -622,7 +551,6 @@ _time_machine_unpatch(PyObject *module, PyObject *unused)
 
     state->date_today->m_ml->ml_meth = state->original_date_today;
     state->original_date_today = NULL;
-    state->original_datetime_today = NULL;
 
     /*
         time.clock_gettime(), only available on Unix platforms.
@@ -673,14 +601,6 @@ static PyMethodDef module_functions[] = {
         (PyCFunction)_time_machine_original_utcnow,
         METH_NOARGS,
         original_utcnow_doc},
-    {"original_date_today",
-        (PyCFunction)_time_machine_original_date_today,
-        METH_NOARGS,
-        original_date_today_doc},
-    {"original_datetime_today",
-        (PyCFunction)_time_machine_original_datetime_today,
-        METH_NOARGS,
-        original_datetime_today_doc},
 #if PY_VERSION_HEX >= 0x030d00a2
     {"original_clock_gettime",
         (PyCFunction)_time_machine_original_clock_gettime,
@@ -752,13 +672,13 @@ _time_machine_exec(PyObject *module)
         goto error;
     }
 
-    state->date_class = PyObject_GetAttrString(state->datetime_module, "date");
-    if (state->date_class == NULL) {
+    PyObject *date_class = PyObject_GetAttrString(state->datetime_module, "date");
+    if (date_class == NULL) {
         goto error;
     }
 
-    state->date_today =
-        (PyCFunctionObject *)PyObject_GetAttrString(state->date_class, "today");
+    state->date_today = (PyCFunctionObject *)PyObject_GetAttrString(date_class, "today");
+    Py_DECREF(date_class);
     if (state->date_today == NULL) {
         goto error;
     }
@@ -826,7 +746,6 @@ _time_machine_exec(PyObject *module)
 error:
     Py_CLEAR(state->datetime_module);
     Py_CLEAR(state->datetime_class);
-    Py_CLEAR(state->date_class);
     Py_CLEAR(state->datetime_datetime_now);
     Py_CLEAR(state->datetime_datetime_utcnow);
     Py_CLEAR(state->date_today);
@@ -847,7 +766,6 @@ _time_machine_traverse(PyObject *module, visitproc visit, void *arg)
     _time_machine_state *state = get_time_machine_state(module);
     Py_VISIT(state->datetime_module);
     Py_VISIT(state->datetime_class);
-    Py_VISIT(state->date_class);
     Py_VISIT(state->datetime_datetime_now);
     Py_VISIT(state->datetime_datetime_utcnow);
     Py_VISIT(state->date_today);
@@ -868,7 +786,6 @@ _time_machine_clear(PyObject *module)
     _time_machine_state *state = get_time_machine_state(module);
     Py_CLEAR(state->datetime_module);
     Py_CLEAR(state->datetime_class);
-    Py_CLEAR(state->date_class);
     Py_CLEAR(state->datetime_datetime_now);
     Py_CLEAR(state->datetime_datetime_utcnow);
     Py_CLEAR(state->date_today);

--- a/src/time_machine/__init__.py
+++ b/src/time_machine/__init__.py
@@ -413,6 +413,16 @@ def utcnow() -> dt.datetime:
     return dt.datetime.fromtimestamp(time(), dt.timezone.utc).replace(tzinfo=None)
 
 
+def date_today() -> dt.date:
+    # Convert timestamp to date in local timezone
+    return dt.datetime.fromtimestamp(time()).date()
+
+
+def datetime_today() -> dt.datetime:
+    # Convert timestamp to datetime in local timezone
+    return dt.datetime.fromtimestamp(time())
+
+
 # time module
 
 
@@ -544,10 +554,21 @@ class _EscapeHatchDatetimeDatetime:
         result: dt.datetime = _time_machine.original_utcnow()
         return result
 
+    def today(self) -> dt.datetime:
+        result: dt.datetime = _time_machine.original_datetime_today()
+        return result
+
+
+class _EscapeHatchDatetimeDate:
+    def today(self) -> dt.date:
+        result: dt.date = _time_machine.original_date_today()
+        return result
+
 
 class _EscapeHatchDatetime:
     def __init__(self) -> None:
         self.datetime = _EscapeHatchDatetimeDatetime()
+        self.date = _EscapeHatchDatetimeDate()
 
 
 class _EscapeHatchTime:

--- a/src/time_machine/__init__.py
+++ b/src/time_machine/__init__.py
@@ -545,30 +545,30 @@ if HAVE_PYTEST:  # pragma: no branch
 # escape hatch
 
 
-class _EscapeHatchDatetimeDatetime:
-    def now(self, tz: dt.tzinfo | None = None) -> dt.datetime:
-        result: dt.datetime = _time_machine.original_now(tz)
-        return result
-
-    def utcnow(self) -> dt.datetime:
-        result: dt.datetime = _time_machine.original_utcnow()
-        return result
-
-    def today(self) -> dt.datetime:
-        result: dt.datetime = _time_machine.original_datetime_today()
-        return result
-
-
 class _EscapeHatchDatetimeDate:
     def today(self) -> dt.date:
         result: dt.date = _time_machine.original_date_today()
         return result
 
 
+class _EscapeHatchDatetimeDatetime:
+    def now(self, tz: dt.tzinfo | None = None) -> dt.datetime:
+        result: dt.datetime = _time_machine.original_now(tz)
+        return result
+
+    def today(self) -> dt.datetime:
+        result: dt.datetime = _time_machine.original_datetime_today()
+        return result
+
+    def utcnow(self) -> dt.datetime:
+        result: dt.datetime = _time_machine.original_utcnow()
+        return result
+
+
 class _EscapeHatchDatetime:
     def __init__(self) -> None:
-        self.datetime = _EscapeHatchDatetimeDatetime()
         self.date = _EscapeHatchDatetimeDate()
+        self.datetime = _EscapeHatchDatetimeDatetime()
 
 
 class _EscapeHatchTime:

--- a/src/time_machine/__init__.py
+++ b/src/time_machine/__init__.py
@@ -413,16 +413,6 @@ def utcnow() -> dt.datetime:
     return dt.datetime.fromtimestamp(time(), dt.timezone.utc).replace(tzinfo=None)
 
 
-def date_today() -> dt.date:
-    # Convert timestamp to date in local timezone
-    return dt.datetime.fromtimestamp(time()).date()
-
-
-def datetime_today() -> dt.datetime:
-    # Convert timestamp to datetime in local timezone
-    return dt.datetime.fromtimestamp(time())
-
-
 # time module
 
 
@@ -547,8 +537,9 @@ if HAVE_PYTEST:  # pragma: no branch
 
 class _EscapeHatchDatetimeDate:
     def today(self) -> dt.date:
-        result: dt.date = _time_machine.original_date_today()
-        return result
+        # date.today() is equivalent to datetime.now().date().
+        result: dt.datetime = _time_machine.original_now(None)
+        return result.date()
 
 
 class _EscapeHatchDatetimeDatetime:
@@ -557,7 +548,8 @@ class _EscapeHatchDatetimeDatetime:
         return result
 
     def today(self) -> dt.datetime:
-        result: dt.datetime = _time_machine.original_datetime_today()
+        # datetime.today() is equivalent to datetime.now() without a timezone.
+        result: dt.datetime = _time_machine.original_now(None)
         return result
 
     def utcnow(self) -> dt.datetime:

--- a/tests/test_time_machine.py
+++ b/tests/test_time_machine.py
@@ -1317,6 +1317,17 @@ class TestEscapeHatch:
         with time_machine.travel(EPOCH):
             assert time_machine.escape_hatch.is_travelling() is True
 
+    def test_date_today(self):
+        real_today = dt.date.today()
+
+        with time_machine.travel(EPOCH):
+            eh_today = time_machine.escape_hatch.datetime.date.today()
+            assert eh_today >= real_today
+
+        with pytest.raises(ValueError) as excinfo:
+            time_machine.escape_hatch.datetime.date.today()
+        assert excinfo.value.args == ("Not currently time-travelling.",)
+
     def test_datetime_now(self):
         real_now = dt.datetime.now()
 
@@ -1334,6 +1345,17 @@ class TestEscapeHatch:
         with time_machine.travel(EPOCH):
             eh_now = time_machine.escape_hatch.datetime.datetime.now(tz=dt.timezone.utc)
             assert eh_now >= real_now
+
+    def test_datetime_today(self):
+        real_today = dt.datetime.today()
+
+        with time_machine.travel(EPOCH):
+            eh_today = time_machine.escape_hatch.datetime.datetime.today()
+            assert eh_today >= real_today
+
+        with pytest.raises(ValueError) as excinfo:
+            time_machine.escape_hatch.datetime.datetime.today()
+        assert excinfo.value.args == ("Not currently time-travelling.",)
 
     def test_datetime_utcnow(self):
         real_now = dt.datetime.utcnow()

